libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java

```diff
@@ -112,10 +113,25 @@ public XContent xContent() {
             return CborXContent.cborXContent;
         }
     };
-
-    public static final MediaTypeParser<XContentType> mediaTypeParser = new MediaTypeParser<>(XContentType.values(),
-        Map.of("application/*", JSON, "application/x-ndjson", JSON));
-
+    /**
+     * A regexp to allow parsing media types. It covers two use cases.
+     * 1. Media type with a version - requires a custom vnd.elasticsearch subtype and a compatible-with parameter
+     * i.e. application/vnd.elasticsearch+json;compatible-with
+     * 2. Media type without a version - for users not using compatible API i.e. application/json
+     */
+     private static final MediaTypeParser<XContentType> mediaTypeParser = new MediaTypeParser.Builder<XContentType>()
+        .withMediaTypeAndParams("application/smile", SMILE, Collections.emptyMap())
+        .withMediaTypeAndParams("application/cbor", CBOR, Collections.emptyMap())
+        .withMediaTypeAndParams("application/json", JSON, Map.of("charset", "UTF-8"))
+        .withMediaTypeAndParams("application/yaml", YAML, Map.of("charset", "UTF-8"))
+        .withMediaTypeAndParams("application/*", JSON, Map.of("charset", "UTF-8"))
+        .withMediaTypeAndParams("application/x-ndjson", JSON, Map.of("charset", "UTF-8"))
+        .withMediaTypeAndParams("application/vnd.elasticsearch+json", JSON, Map.of("compatible-with", "\\d+","charset", "UTF-8"))
+        .withMediaTypeAndParams("application/vnd.elasticsearch+smile", SMILE, Map.of("compatible-with", "\\d+","charset", "UTF-8"))
+        .withMediaTypeAndParams("application/vnd.elasticsearch+yaml", YAML, Map.of("compatible-with", "\\d+","charset", "UTF-8"))
+        .withMediaTypeAndParams("application/vnd.elasticsearch+cbor", CBOR, Map.of("compatible-with", "\\d+","charset", "UTF-8"))
+        .withMediaTypeAndParams("application/vnd.elasticsearch+x-ndjson", JSON, Map.of("compatible-with", "\\d+","charset", "UTF-8"))
+        .build();
```

Comment:
defining versioned media types up front, means that we won't throw a exception if someone specifies it with oss licence
if he uses oss on v8 server for api that was removed and provides `application/vnd.elasticsearch+json;compatible-with=7` he will get a 404

is there an easy way to make `XContentType` plugin aware? or distro aware?

Reply:
but this also would affect plugins like SQL. In oss we don't know about MediaTypes which are defined over there
if we throw an exception for a unknown type I am worried we might end up with a logic based on exceptions
```
try XContentType.parseMediaType
try XContentType.parseFormat
try TextFormat.parseMediaType
try TextFormat.parseFormat
```

@bpintea any thoughts on this? I guess SQL had to face this in the past

Reply:
Is there any way to register a custom parser ? or just add the SQL values to the core parser and specificy which rule set to use when parsing ?

Reply:
that is an interesting idea.. We could have something like a SqlMediaTypeParser and it would either return TextFormat or XContentType. The parsing logic would remain the same in the MediaTypeParser. The idea of SqlMediaTypeParser woudl be just to abstract on the setup of the parser.
The same could be done for CompatibleApiMediaTypeParser and we could keep it in xpack plugin.
this does not solve the problem of trying to parse a format, then if failed to parse a media type. I think we should try to just check for presence of the format parameter and if present to used. Otherwise trying to use Accept Header.
I think also that Content-Type should not be used for response formatting in SQL plugin. It looks like it is allowed in code, but I don't think it would work (it would fail when parsing that header in server)